src/expr/src/expr/expr_concat_ws.rs

```diff
@@ -0,0 +1,177 @@
+// Copyright 2022 Singularity Data
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::convert::TryFrom;
+use std::sync::Arc;
+
+use risingwave_common::array::{
+    Array, ArrayBuilder, ArrayImpl, ArrayRef, DataChunk, Utf8ArrayBuilder,
+};
+use risingwave_common::error::{Result, RwError};
+use risingwave_common::types::DataType;
+use risingwave_common::{ensure, try_match_expand};
+use risingwave_pb::expr::expr_node::{RexNode, Type};
+use risingwave_pb::expr::ExprNode;
+
+use crate::expr::{build_from_prost as expr_build_from_prost, BoxedExpression, Expression};
+
+#[derive(Debug)]
+pub struct ConcatWsExpression {
+    return_type: DataType,
+    sep_expr: BoxedExpression,
+    string_exprs: Vec<BoxedExpression>,
+}
+
+impl Expression for ConcatWsExpression {
+    fn return_type(&self) -> DataType {
+        self.return_type.clone()
+    }
+
+    fn eval(&self, input: &DataChunk) -> Result<ArrayRef> {
+        let sep_column = self.sep_expr.eval(input)?;
+        let sep_column = sep_column.as_utf8();
+
+        let string_columns = self
+            .string_exprs
+            .iter()
+            .map(|c| c.eval(input))
+            .collect::<Result<Vec<_>>>()?;
+
+        let mut builder = Utf8ArrayBuilder::new(input.cardinality())?;
+        let row_len = string_columns[0].len();
+
+        for row_idx in 0..row_len {
+            let sep = sep_column.value_at(row_idx);
+            if sep.is_none() {
+                builder.append(None)?;
+                continue;
+            }
+            let sep = sep.unwrap();
+
+            let mut res = String::from("");
+            let mut found_first = false;
+
+            for string_column in &string_columns {
+                let string_column = string_column.as_utf8();
+                if let Some(string) = string_column.value_at(row_idx) {
+                    if !found_first {
```

Comment:
we can also move this `found_first = true` and `res.push_str(string);` out of the `for string_column` loop to eliminate the if branch.